Sources/NIO/Selector.swift

```diff
@@ -41,6 +41,192 @@ private extension Optional {
+struct SelectorEventSet: OptionSet, Equatable {
+
+    typealias RawValue = UInt8
+
+    let rawValue: UInt8
+
+    /// It's impossible to actually register for no events, therefore `_none` should only be used to bootstrap a set
+    /// of flags or to compare against spurious wakeups.
+    static let _none = SelectorEventSet(rawValue: 0)
+
+    /// Connection reset or other errors.
+    static let reset = SelectorEventSet(rawValue: 1 << 0)
+
+    /// EOF at the read/input end of a `Selectable`.
+    static let readEOF = SelectorEventSet(rawValue: 1 << 1)
+
+    /// Interest in/availability of data to be read
+    static let read = SelectorEventSet(rawValue: 1 << 2)
+
+    /// Interest in/availability of data to be written
+    static let write = SelectorEventSet(rawValue: 1 << 3)
+
+    init(rawValue: SelectorEventSet.RawValue) {
+        self.rawValue = rawValue
+    }
+}
+
+/// Represents the `kqueue` filters we might use:
+///
+///  - `except` corresponds to `EVFILT_EXCEPT`
+///  - `read` corresponds to `EVFILT_READ`
+///  - `write` corresponds to `EVFILT_WRITE`
+private struct KQueueEventFilterSet: OptionSet, Equatable {
+    typealias RawValue = UInt8
+
+    let rawValue: UInt8
+
+    static let _none = KQueueEventFilterSet(rawValue: 0)
+    static let except = KQueueEventFilterSet(rawValue: 1 << 1)
+    static let read = KQueueEventFilterSet(rawValue: 1 << 2)
+    static let write = KQueueEventFilterSet(rawValue: 1 << 3)
+
+    init(rawValue: RawValue) {
+        self.rawValue = rawValue
+    }
+}
+
+/// Represents the `epoll` filters/events we might use:
+///
+///  - `hangup` corresponds to `EPOLLHUP`
+///  - `readHangup` corresponds to `EPOLLRDHUP`
+///  - `input` corresponds to `EPOLLIN`
+///  - `output` corresponds to `EPOLLOUT`
+///  - `error` corresponds to `EPOLLERR`
+private struct EpollFilterSet: OptionSet, Equatable {
+    typealias RawValue = UInt8
+
+    let rawValue: UInt8
+
+    static let _none = EpollFilterSet(rawValue: 0)
+    static let hangup = EpollFilterSet(rawValue: 1 << 0)
+    static let readHangup = EpollFilterSet(rawValue: 1 << 1)
+    static let input = EpollFilterSet(rawValue: 1 << 2)
+    static let output = EpollFilterSet(rawValue: 1 << 3)
+    static let error = EpollFilterSet(rawValue: 1 << 4)
+
+    init(rawValue: RawValue) {
+        self.rawValue = rawValue
+    }
+}
+
+extension KQueueEventFilterSet {
+    /// Convert NIO's `SelectorEventSet` set to a `KQueueEventFilterSet`
+    init(selectorEventSet: SelectorEventSet) {
+        var thing: KQueueEventFilterSet = .init(rawValue: 0)
+        if selectorEventSet.contains(.read) {
+            thing.formUnion(.read)
+        }
+        if selectorEventSet.contains(.write) {
+            thing.formUnion(.write)
+        }
+
+        // kqueue gives us a `EV_EOF` for the input automatically with `EVFILT_READ` so there's no need to register
+        // for both `EVFILT_EXCEPT` and `EVFILT_READ`.
+
+        // FIXME: this needs to also have .reset
```

Comment:
Is this FIXME actually addressed already?

Reply:
no, removed it. We said we'll do a follow-up to emulate `EPOLLHUP` with kqueue using `EVFILT_WRITE` and `ONESHOT`.